istioctl/cmd/wait.go

```diff
@@ -0,0 +1,206 @@
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/api/meta"
+
+	"istio.io/istio/istioctl/pkg/kubernetes"
+	v2 "istio.io/istio/pilot/pkg/proxy/envoy/v2"
+)
+
+var (
+	forDistribution bool
+	targetResource  string
+	forDelete       bool
+	threshold       float32
+	timeout         time.Duration
+	resourceVersion string
+	//resourceVersions []string
+	rvLock sync.Mutex
+)
+
+const pollInterval = time.Second
+
+// waitCmd represents the wait command
+func waitCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "wait",
+		Short: "Wait for an Istio Resource",
+		Long: `A longer description that spans multiple lines and likely contains examples
+and usage of using your command. For example:
+
+Cobra is a CLI library for Go that empowers applications.
+This application is a tool to generate the needed files
+to quickly create a Cobra application.`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var versionChan chan string
+			var g *errgroup.Group
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			defer cancel()
+			if resourceVersion == "" {
+				versionChan, g = getAndWatchResource(ctx, targetResource) // setup version getter from kubernetes
+			} else {
+				versionChan = make(chan string, 1)
+				versionChan <- resourceVersion
+			}
+			// wait for all deployed versions to be contained in resourceVersions
+			t := time.NewTicker(pollInterval)
+			resourceVersions := make([]string, 1)
+			for {
+				select {
+				case newVersion := <-versionChan:
+					resourceVersions = append(resourceVersions, newVersion)
+				case <-t.C:
+					finished, err := poll(resourceVersions)
+					if err != nil {
+						return err
+					} else if finished {
+						return nil
+					}
+				case <-ctx.Done():
+					// I think this means the timeout has happened:
+					t.Stop()
+					if err := g.Wait(); err != nil {
+						return err
+					}
+					return errors.New("timeout expired before resource was distributed")
+				}
+			}
+		},
+		Args: func(cmd *cobra.Command, args []string) error {
+			if err := cobra.ExactArgs(1)(cmd, args); err != nil {
+				return err
+			}
+			targetResource = args[0]
+			return parseResource(targetResource)
+		},
+	}
+	cmd.PersistentFlags().BoolVar(&forDistribution, "for-distribution", false,
+		"wait for the designated resource to be distributed")
+	cmd.PersistentFlags().BoolVar(&forDelete, "for-delete", false,
+		"wait for the designated resource to be distributed")
+	cmd.PersistentFlags().DurationVar(&timeout, "timeout", time.Second*30,
+		"the duration to wait before failing (default 30s)")
+	cmd.PersistentFlags().Float32Var(&threshold, "threshold", 1,
+		"The ratio of distribution required for success (default 1.0)")
+	cmd.PersistentFlags().StringVar(&resourceVersion, "resource-version", "",
+		"Causes istioctl to wait for a specific version of config to become current, rather than using whatever is latest in k8s")
```

Comment:
`Kubernetes` instead of `k8s`.  Follow the documentation conventions for any messages that might appear to a user.  https://istio.io/about/contribute/style-guide/ .  I think we can also eliminate "Causes istioctl to " at the  the beginning.